### PR TITLE
RFC add markdown Table

### PR DESCRIPTION
**Pull request by hayd**
This parses and displays tables in Markdown with [GH syntax](https://help.github.com/articles/github-flavored-markdown#tables), in the terminal and html. _I haven't tested the latex version and the plain is not yet finished._

As discussed in other documentation issues, tables are useful to have...

cc @one-more-minute, fixes https://github.com/one-more-minute/Markdown.jl/issues/16.

No doubt there are a few things I'm not doing correctly here, so comments would be great.

---

~~Note: There aren't any markdown tests at all...~~

### CODE REVIEW

**Comment by one-more-minute** [comment]
Nice! Great to see this.

After a skim of the code it looks pretty reasonable, but I'll need to play around with it a bit. I'll try to get that done within a few days.

File: base/markdown/render/plain.jl [hunk 1]
```diff
@@ -37,6 +37,46 @@ function plain(io::IO, list::List)
...
+    end
+end
+
+function _dash(width, align)
+    if align == :l
+        return ":" * "-" ^ max(1, width - 1)
+    elseif align == :r
+        return "-" ^ max(1, width - 1) * ":"
+    elseif align == :c
+        return "-" ^ width
```
Comment (hayd) [inline base/markdown/render/plain.jl]: Hmm this should be `":" * "-" ^ max(1, width - 2) * ":"`, since default is to right align.

**Comment by hayd** [comment]
@one-more-minute thanks for looking, I just pushed a commit to simplify (IOBuffer with SubString) but in retrospect it doesn't work, will see if I can fix it, or will just delete that commit. Sorry!

File: base/markdown/GitHub/GitHub.jl [hunk 2]
```diff
@@ -33,7 +33,83 @@ function github_paragraph(stream::IO, md::MD, config::Config)
...
+        push!(md, Table(rows, align))
+        return true
+    end
+end
+
+_full{T}(s::SubString{T}) = convert(T, s)
+_full(s::AbstractString) = s
+
+@flavor github [list, indentcode, blockquote, fencedcode, hashheader,
+                github_paragraph, github_table,
```
Comment (hayd) [inline base/markdown/GitHub/GitHub.jl]: it may be that table needs to be earlier in this list, is this basically precedence?

Comment (one-more-minute) [inline base/markdown/GitHub/GitHub.jl]: Yup, it's precedence. If the table could be interpreted as another block element (such as a paragraph) it needs to go before that element. I imagine that swapping table and paragraph here would be the right thing. You'll want to also have the same order in the `julia` flavour.

Comment (hayd) [inline base/markdown/GitHub/GitHub.jl]: I put it first in the julia `@flavor` and it seemed to work, will update (I think being before paragraph is the important thing).

I suspect there may be some edge cases - are precedence for markdown parsing published somewhere? We really need some tests here IMO (will put a few together in another PR)...

**Comment by hayd** [comment]
@one-more-minute updated precedence and fixed the plain printing. I've not checked latex, not sure how to call it.

**Comment by one-more-minute** [comment]
RE tests, feel free to add a Markdown tests file if you want. I'll try and get equality working to make this easier.

FWIW, GitHub's Markdown parser appears to be a bit stricter than this code. At least three dashes are required, and if you're using an outer pipe it mustn't be indented (so the example below wouldn't parse). I'm not strongly opinionated about what the "right" way here is, so if you're happy with the current behaviour I'm cool with that, but in might be worth bearing in mind.

Inline formatting throws an error when rendering, currently:
```julia
julia> md"""
       | a | b | c |
         | - | - | - |
       | *d* | e | f |
       """
Error showing value of type Base.Markdown.MD:
ERROR: MethodError: `replace` has no method matching replace(::Base.Markdown.Italic, ::Regex, ::ASCIIString)
 in ansi_length at markdown/render/terminal/formatting.jl:48
 in ansi_length at markdown/render/terminal/render.jl:87
```
I'm guessing you just need to render to a string via `plaininline`/`terminline` before calling `ansilength`.

I'll take a look over the code itself and add any other comments I can think of.

@MichaelHatherly any chance you could have a quick look over the LaTeX code?

File: base/markdown/GitHub/GitHub.jl [hunk 3]
```diff
@@ -33,7 +33,83 @@ function github_paragraph(stream::IO, md::MD, config::Config)
     return true
 end
 
-# TODO: tables
+typealias Row Vector
```
Comment (one-more-minute) [inline base/markdown/GitHub/GitHub.jl]: I think it'd be best to make this explicitly `Vector{Any}`, or you risk running into subtle typing errors.

File: base/markdown/GitHub/GitHub.jl [hunk 4]
```diff
@@ -33,7 +33,83 @@ function github_paragraph(stream::IO, md::MD, config::Config)
...
+                    end
+                end
+
+            elseif n == 1 || length(rows[1]) == length(row)
+                push!(rows, Row(map(x -> parseinline(IOBuffer(_full(x)), config), row)))
+            elseif length(row) > 1
+                seek(stream, pos)
+                break
+            else
+                seek(stream, 0)
```
Comment (one-more-minute) [inline base/markdown/GitHub/GitHub.jl]: `seek(stream,0)` definitely isn't what you want to do here. You can just delete these lines because `return false` will handle seeking the stream for you.

File: base/markdown/GitHub/GitHub.jl [hunk 5]
```diff
@@ -33,7 +33,83 @@ function github_paragraph(stream::IO, md::MD, config::Config)
...
+end
+
+function github_table(stream::IO, md::MD, config::Config)
+    withstream(stream) do
+        rows = Row[]
+        n = 0
+        align = :r # default is to align right
+        while !eof(stream)
+            n += 1
+            pos = position(stream)
```
Comment (one-more-minute) [inline base/markdown/GitHub/GitHub.jl]: Since you're repeating the `withstream` pattern here by hand, I'm thinking it might be best to factor row parsing into its own `parse_row` function, or something similar. e.g.
```julia
function parse_row(stream::IO)
  withstream(stream) do
    # return row vector or `nothing`, without explicitly seeking the stream
```
Then just call `parse_row` until it returns `nothing`.

File: base/markdown/Julia/Julia.jl [hunk 6]
```diff
@@ -7,7 +7,7 @@ We start by borrowing GitHub's `fencedcode` extension – more to follow.
 
 include("interp.jl")
 
-@flavor julia [blocktex, blockinterp, hashheader, list, indentcode, fencedcode,
+@flavor julia [github_table, blocktex, blockinterp, hashheader, list, indentcode, fencedcode,
```
Comment (one-more-minute) [inline base/markdown/Julia/Julia.jl]: I don't think tables will cause conflicts with other block elements, but performance-wise it's probably best to run it as late as possible. For that reason I'd put it just before `paragraph`.

File: base/markdown/render/html.jl [hunk 7]
```diff
@@ -56,6 +56,21 @@ function html(io::IO, md::List)
     end
 end
 
+function html(io::IO, md::Table)
+    withtag(io, :table) do
+        for (i, row) in enumerate(md.rows)
+            withtag(io, :tr) do
+                for c in md.rows[i]
+                    t = i == 1 ? :th : :td
```
Comment (one-more-minute) [inline base/markdown/render/html.jl]: Very minor thing, but I find this a little hard to read without brackets around the ternary operator.

File: base/markdown/render/plain.jl [hunk 8]
```diff
@@ -37,6 +37,46 @@ function plain(io::IO, list::List)
     end
 end
 
+function plain(io::IO, md::Table)
+    col_widths = reduce(max, map(cell -> map(ansi_length, cell), md.rows))
```
Comment (one-more-minute) [inline base/markdown/render/plain.jl]: If the plain text has ansi codes in, something's gone wrong ;) `length` should be fine here, and you'll want to render each cell beforehand as well.

File: base/markdown/render/plain.jl [hunk 9]
```diff
@@ -37,6 +37,46 @@ function plain(io::IO, list::List)
...
+    for (n, row) in enumerate(md.rows)
+        for (i, h) in enumerate(row)
+            # TODO use not terminal version of print_align
+            if i != 1
+                print(io, " | ")
+            end
+            print(io, " " ^ (col_widths[i] - ansi_length(h)))
+            plaininline(io, h)
+        end
+        println(io, "")
```
Comment (one-more-minute) [inline base/markdown/render/plain.jl]: `println(io)` will do that for you

File: base/markdown/render/plain.jl [hunk 10]
```diff
@@ -37,6 +37,46 @@ function plain(io::IO, list::List)
     end
 end
 
+function plain(io::IO, md::Table)
+    col_widths = reduce(max, map(cell -> map(ansi_length, cell), md.rows))
+
+    for (n, row) in enumerate(md.rows)
+        for (i, h) in enumerate(row)
+            # TODO use not terminal version of print_align
+            if i != 1
```
Comment (one-more-minute) [inline base/markdown/render/plain.jl]: It's quite common in Julia to write things like this as `i == 1 || print(io, " | ")`.

File: base/markdown/GitHub/GitHub.jl [hunk 11]
```diff
@@ -33,7 +33,83 @@ function github_paragraph(stream::IO, md::MD, config::Config)
...
+type Table
+    rows::Vector{Row}
+    align
+end
+
+function github_table(stream::IO, md::MD, config::Config)
+    withstream(stream) do
+        rows = Row[]
+        n = 0
+        align = :r # default is to align right
```
Comment (one-more-minute) [inline base/markdown/GitHub/GitHub.jl]: This looks a bit strange to me:
```julia
  a b c 
--- - - 
foo e f 
```
Maybe it would look better if the alignment didn't apply to headers?

File: base/markdown/render/plain.jl [hunk 12]
```diff
@@ -37,6 +37,46 @@ function plain(io::IO, list::List)
...
+            println(io, "")
+        end
+    end
+end
+
+function _dash(width, align)
+    if align == :l
+        return ":" * "-" ^ max(1, width - 1)
+    elseif align == :r
+        return "-" ^ max(1, width - 1) * ":"
```
Comment (one-more-minute) [inline base/markdown/render/plain.jl]: It might be nicer not to print it in this case, since it's the default anyway.

File: base/markdown/GitHub/GitHub.jl [hunk 13]
```diff
@@ -33,7 +33,83 @@ function github_paragraph(stream::IO, md::MD, config::Config)
...
+                            push!(align, :r)
+                        else
+                            # default is align right
+                            push!(align, :r)
+                        end
+                    end
+                end
+
+            elseif n == 1 || length(rows[1]) == length(row)
+                push!(rows, Row(map(x -> parseinline(IOBuffer(_full(x)), config), row)))
```
Comment (one-more-minute) [inline base/markdown/GitHub/GitHub.jl]: You shouldn't need the `IOBuffer` constructor here, since `parseinline` has backup methods for strings. You might be able to get away with the `_full`, too, and if not I think it should be `parseinline`'s job to do the conversion.

File: base/markdown/render/plain.jl (hunk 8 above)
Comment (hayd) [inline base/markdown/render/plain.jl]: So the issue here is that length is the length of the array of components rather than the total length of the strings. I could define length too for these - I didn't know the purpose of plain tbh (is it a neat/aligned representation of the tables or just one that is valid markdown - that can be parsed back)?

File: base/markdown/GitHub/GitHub.jl (hunk 11 above)
Comment (hayd) [inline base/markdown/GitHub/GitHub.jl]: That's kinda an edge case :). I guess I'm used to pandas and right alignment seems normal for me. Perhaps two spaces between columns would make this clearer?

File: base/markdown/render/plain.jl (hunk 12 above)
Comment (hayd) [inline base/markdown/render/plain.jl]: I had this in just in case we changed the default. I did wonder if this could (later) be configurable i.e. originally I had a :default symbol but for simplicity just changed to the latex way.

File: base/markdown/render/plain.jl (hunk 8 above)
Comment (one-more-minute) [inline base/markdown/render/plain.jl]: Yeah, the idea is for it to give you Markdown back.

Using `ansi_length` doesn't really solve that problem, since it only operates on strings, and not inline markdown elements such as italic. It'd be much easier to render each cell to a string via `plaininline` and then call the usual `length` function.

File: base/markdown/GitHub/GitHub.jl (hunk 13 above)
Comment (hayd) [inline base/markdown/GitHub/GitHub.jl]: IIRC I didn't work without the `_full`, will have a go! A lot of this is copy and paste from paragraph btw :p

File: base/markdown/GitHub/GitHub.jl (hunk 4 above)
Comment (hayd) [inline base/markdown/GitHub/GitHub.jl]: IIRC paragraph does this too,  will see what happens!

File: base/markdown/render/terminal/render.jl [hunk 14]
```diff
@@ -61,6 +61,36 @@ function term(io::IO, md::Code, columns)
...
+        if n == 1
+            for w in col_widths
+                print(io, ("-" ^ w) * " ")
+            end
+            println(io, "")
+        end
+    end
+end
+
+function ansi_length(md::Vector{Any})
```
Comment (one-more-minute) [inline base/markdown/render/terminal/render.jl]: I don't really like this overloading – `ansi_length` has a very narrow purpose ("length of a string minus ansi characters") and you're overloading it to mean "string length of a markdown document". The best way to get the string length of some `md` is `length(terminline(md))`.

File: base/markdown/GitHub/GitHub.jl (hunk 11 above)
Comment (one-more-minute) [inline base/markdown/GitHub/GitHub.jl]: Two spaces or an explicit delimiter (`|`) might be clearer. But I'm not too worried – we can try it and see if people complain ;)

File: base/markdown/GitHub/GitHub.jl (hunk 13 above)
Comment (one-more-minute) [inline base/markdown/GitHub/GitHub.jl]: No problem :P If it doesn't work without `_full` I recommend adding a new method to `parseinline` that takes a substring and converts it to a full string, instead of having a `_full` function.

File: base/markdown/render/terminal/render.jl (hunk 14 above)
Comment (hayd) [inline base/markdown/render/terminal/render.jl]: excellent, thanks! Hopefully this fixes the first breaking example (with italics) that you posted.

File: base/markdown/GitHub/GitHub.jl (hunk 11 above)
Comment (hayd) [inline base/markdown/GitHub/GitHub.jl]: Let's go with two spaces for now and if people complain we can do something more elaborate. e.g.

```
+----+-----+
|col | col2|
+----+-----+
|1   |2    |
+----+-----+
```

One thing that I wanted initially was to have color (yellow), however one formatting e.g. code obliterates the formatting, you can see this with 

    md"# Bold `blue` no longer bold"

Comment (one-more-minute) [inline base/markdown/GitHub/GitHub.jl]: Using colours is a good idea. I think that can probably fixed in the terminal rendering code so I'll take a look at that.

Comment (hayd) [inline base/markdown/GitHub/GitHub.jl]: Have addressed this with column width being minimum of 3.

File: base/markdown/GitHub/GitHub.jl (hunk 5 above)
Comment (hayd) [inline base/markdown/GitHub/GitHub.jl]: Can we refactor this part once we have some tests, after this is merged? (And we have some tests). I think I've addressed your other points in: https://github.com/hayd/julia/commit/ee27d425bfc56e15abd2a31e9b24adb204a97693 Thanks!

File: base/markdown/render/terminal/render.jl (hunk 14 above)
Comment (hayd) [inline base/markdown/render/terminal/render.jl]: Apparently I missed this issue. Just to comment, FYI, that `length(terminline(md))` didn't work because it doesn't skip the ansi codes, unfortunately plaininline had a side effect, so I modified it to just return the string. Not sure how to get rid of this call elegantly (since it needs to work recursively).

Comment (hayd) [inline base/markdown/render/terminal/render.jl]: I think it's worth noting that this method is not exported.

Comment (one-more-minute) [inline base/markdown/render/terminal/render.jl]: Sorry, I've mixed the two things up, but you'll want `ansi_length(terminline(md))` or `length(plaininline(md))`. That's probably not quite literally what you want, though – if you're going to render the content it's best to save the result and use it again later.

File: base/markdown/render/plain.jl [hunk 15]
```diff
@@ -47,20 +86,26 @@ function plaininline(io::IO, md...)
...
-plaininline(io::IO, md::Italic) = plaininline(io, "*", md.text, "*")
+plaininline(md::Bold) = "**" * plaininline(md.text) * "**"
 
-plaininline(io::IO, md::Code) = print(io, "`", md.code, "`")
+plaininline(md::Italic) = "*" * plaininline(md.text) + "*"
 
+plaininline(md::Code) = "`$(md.code)`"
+
+plaininline(io::IO, md::Vector) = plaininline(io, md...)
+plaininline(io::IO, md::Union(Image, String, Bold, Italic, Code, Link, Image)) = print(plaininline(md))
```
Comment (one-more-minute) [inline base/markdown/render/plain.jl]: Constructing the string is going to be slow here, so it's best to do it the other way around – leave the `plaininline(io...` methods as they are and have a `plaininline(x) = sprint(plaininline, x)` method if it's not there already.

File: base/markdown/render/plain.jl [hunk 16]
```diff
@@ -37,6 +37,45 @@ function plain(io::IO, list::List)
     end
 end
 
+function plain(io::IO, md::Table)
+    col_widths = reduce(max, map(row -> int(map(x -> length(plaininline(x)), row)), md.rows))
```
Comment (one-more-minute) [inline base/markdown/render/plain.jl]: Right idea, but note that `plaininline` is potentially expensive, so calling it three times redundantly isn't optimal. If I were you I'd map the list of cells to a list of rendered cells (via `plaininline`), then work directly with that for the rest of the function.

File: base/markdown/GitHub/GitHub.jl (hunk 5 above)
Comment (one-more-minute) [inline base/markdown/GitHub/GitHub.jl]: I've added some Markdown tests [here](https://github.com/JuliaLang/julia/blob/master/test/markdown.jl), so feel free to add tests as part of this PR.

**Comment by one-more-minute** [comment]
Ok, this is coming along nicely, a little bit of polish and we should be good to go.

I've improved the way terminal formatting is done, so if you rebase this PR against master you should be able to use prettier output without any trouble (but let me know if you come across anything).

The other thing is: Since tables are fairly complex elements that aren't part of standard markdown, I think it would be best not to mix this code with the other elements. Let's have a `table.jl` file (with all the table parsing and rendering code) next to `github.jl` and include it there.

**Comment by hayd** [comment]
thanks for the feedback, that all sound good.

> Let's have a table.jl file (with all the table parsing and rendering code) 

Just to clarify, extract parsing to Github/table.jl, i think i should make a new file for rendering too in render/table.jl (this seems convention, and as in future Table may not be gh specific) ?

**Comment by one-more-minute** [comment]
I meant put everything in GitHub/table.jl. It's only a hundred lines or so, but it's complex enough that I think making it really self-contained will be much easier to follow.

**Comment by hayd** [comment]
@one-more-minute moved to table.jl, have a final refactor of plain (as you suggested). Adding some tests and I came across the following strange behaviour, any ideas on how to fix? (does it matter?)

```
julia> m = Markdown.parse("""a|b
       1|2""")
  a    b
---  ---
  1    2

julia> 1, m
(1,Markdown.MD(Any[Markdown.Table([Any[Any["a"],Any["b"]],Any[Any["1"],Any["2"]]],:r)],Dict{Any,Any}()))

julia> Markdown.MD(Any[Markdown.Table([Any[Any["a"],Any["b"]],Any[Any["1"],Any["2"]]],:r)],Dict{Any,Any}())
  'a'
---
  'b'
  '1'
  '2'

Note: it works fine if you prepend the first array with Any

julia> Markdown.MD(Any[Markdown.Table(Any[Any[Any["a"],Any["b"]],Any[Any["1"],Any["2"]]],:r)],Dict{Any,Any}())
  a    b
---  ---
  1    2
```

---

@MichaelHatherly latex is commented out in the `Markdown.jl` file, it seems to have missed a refactor after Block/Inline (not sure if the outcome of https://github.com/one-more-minute/Markdown.jl/pull/5 is relevant), IMO inclusion of Inline, Block and Content types would have been useful in the last few changes I've been doing (even if they are just Unions), is there a good reason to ditch them @one-more-minute ?

File: test/markdown.jl [hunk 17]
```diff
@@ -39,3 +39,17 @@ writemime(io::IO, m::MIME"text/plain", r::Reference) =
...
+1|2""" == MD(Table(Any[Any[Any["a"],Any["b"]];Any[Any["1"],Any["2"]]], :r))
+
+table = md"""| a  |  b | c |
+             | :- | -: | - |
+             | d`gh`hg | hgh**jhj**ge | f |"""
+@test table == MD(Table(Any[Any[Any["a"],Any["b"],Any["c"]],
+                            Any[Any["d",Markdown.Code("","gh"),"hg"],
+                                Any["hgh",Markdown.Bold(Any["jhj"]),"ge"],
+                                Any["f"]]],
+                        [:l, :r, :r]))
```
Comment (one-more-minute) [inline test/markdown.jl]: It would be nice to make the tests as human-readable as possible, e.g. you can write this as
```julia
MD(Table(Any[["a","b","c"],
             Any[["d",Code("gh"),"hg"],
                 ["hgh",Bold("jhj"),"ge"],
                 "f"]],
         [:l, :r, :r]))
```
(This actually doesn't work at the moment, which I think is due to a problem with `print_align`)

File: base/markdown/render/terminal/formatting.jl [hunk 18]
```diff
@@ -97,6 +97,25 @@ function print_centred(io::IO, s...; columns = 80, width = columns)
...
+        lpad = width - n
+    elseif align == :l # left
+        lpad = 0
+    else
+        throw(ArgumentError("Alignment $align not recognised"))
+    end
+    print(io, " " ^ lpad)
+    for t in text
+        terminline(io, t)
+    end
```
Comment (one-more-minute) [inline base/markdown/render/terminal/formatting.jl]: There's no need to depend on `text` being an array here – `terminline(io, t)` will be fine

File: base/markdown/render/plain.jl [hunk 19]
```diff
@@ -51,19 +51,19 @@ function plaininline(io::IO, md...)
...
-plaininline(io::IO, md::Bold) = plaininline(io, "**", md.text, "**")
-
-plaininline(io::IO, md::Italic) = plaininline(io, "*", md.text, "*")
-
+plaininline(io::IO, md::Image) = print(io, "![", plaininline(md.alt), "](",
+                                       md.url, ")")
+plaininline(io::IO, md::Link) = print(io, "[", plaininline(md.text), "](",
+                                      md.url, ")")
+plaininline(io::IO, md::Bold) = print(io, "**", plaininline(md.text), "**")
+plaininline(io::IO, md::Italic) = print(io, "*", plaininline(md.text), "*")
```
Comment (one-more-minute) [inline base/markdown/render/plain.jl]: Good to see the extra methods here but there are a lot of unnecessary stylistic changes – would be great to see the diff being as simple as possible

File: base/markdown/render/plain.jl [hunk 20]
```diff
@@ -51,19 +51,19 @@ function plaininline(io::IO, md...)
...
 plaininline(io::IO, md::Code) = print(io, "`", md.code, "`")
-
+plaininline(io::IO, s::String) = print(io, s)
 plaininline(io::IO, x) = writemime(io, MIME"text/plain"(), x)
 
+plaininline(s::String) = s
```
Comment (one-more-minute) [inline base/markdown/render/plain.jl]: This method should be unnecessary, it's covered by the one below

File: base/markdown/GitHub/table.jl [hunk 21]
```diff
@@ -0,0 +1,168 @@
...
+            println(io, " \\\\")
+            if i == 1
+                println("\\hline")
+            end
+        end
+    end
+end
+
+function term(io::IO, md::Table, columns)
+    col_widths = reduce(max, map(cell -> int(map(ansi_length, cell)),
```
Comment (one-more-minute) [inline base/markdown/GitHub/table.jl]: The overloading of `ansi_length` is still problematic. For one thing, see what happens if you use the custom `ref` type defined in the tests in a cell. It would be a lot more robust to get rid of the overloading and use the same approach as `plain` (i.e. render the cells to strings and then call `ansi_length`).

File: base/markdown/GitHub/table.jl [hunk 22]
```diff
@@ -0,0 +1,168 @@
...
+                            # default is align right
+                            push!(align, :r)
+                        end
+                    end
+                end
+
+            elseif n == 1 || length(rows[1]) == length(row)
+                push!(rows, map(x -> parseinline(x, config), row))
+            elseif length(row) > 1
+                seek(stream, pos)
```
Comment (one-more-minute) [inline base/markdown/GitHub/table.jl]: There's a strong practice in Markdown.jl of not handling the stream position explicitly. I guess it's not essential, but it would still be nice to see row parsing factored out to address that

**Comment by one-more-minute** [comment]
The reason you sometimes need `Any[...` is that `[...]` concatenates by default, whereas `T[...]` doesn't. e.g.
```julia
x = [1,2,3]
y = [4,5,6]
[x, y]
[x y]
```
This behaviour is fairly likely to change in future, I think.

File: base/markdown/render/plain.jl (hunk 20 above)
Comment (hayd) [inline base/markdown/render/plain.jl]: This just saves copying a string.

File: base/markdown/GitHub/table.jl (hunk 22 above)
Comment (hayd) [inline base/markdown/GitHub/table.jl]: It may be easier to refactor some of this after merging

File: base/markdown/render/plain.jl (hunk 20 above)
Comment (one-more-minute) [inline base/markdown/render/plain.jl]: Strings are stored in generated markdown trees as `Any["foo"]` though, so this method will almost never be called – kind of a premature optimisation. Not a big deal either way though I guess.

File: base/markdown/GitHub/table.jl (hunk 21 above)
Comment (hayd) [inline base/markdown/GitHub/table.jl]: the problem is that this is not equivalent, plain includes things like `*` or `[]()`, similarly if we were to capture the print it would include ansi codes.

I think we can refactor after merging... :)

Comment (one-more-minute) [inline base/markdown/GitHub/table.jl]: Sure, I don't mean render it via `plaininline`. You'd render it via `terminline` then take the `ansi_length` of that string. It's the exact same approach taken by the wrapping print functions, for example, and anything else that needs the length of the rendered content. There's no other way to do it generally.

Comment (hayd) [inline base/markdown/GitHub/table.jl]: Ah, I see ansi_length *of that string* (from terminline), that makes sense! Will see if this means can delete ansi_length for all the other types.

Would be good if we could test term for these edge cases (e.g. Ref in a table).

**Comment by hayd** [comment]
@one-more-minute I think I've addressed you comments for this and for the sections PR. Would be great to get these merged! :)

*I want to import, and fix up, latex (currently not imported at all) and test it another PR, as well as some other stuff, but am waiting on these two.*

**Comment by hayd** [comment]
bump!

**Comment by ViralBShah** [comment]
Could you squash the commits here too?

**Comment by one-more-minute** [comment]
@hayd Ok, I finally have some time today – let me know when you've squashed, I'll merge and do some of the refactoring.

**Comment by one-more-minute** [comment]
Actually, since you're probably asleep I'll do some rebasing for you, if that's OK. That way you won't have to wait until next weekend again ;)
